### PR TITLE
Add home page

### LINKED ISSUE #104: Implement Homepage

### PR DESCRIPTION
**Pull request by pnicto**
closes #104

Things that need further input:
- [ ] To `useCallback` or not to. (Will require some refactoring to use the hook on the filter function)
- [ ] Page navigation (includes the navigation on card click assuming we are going with the entire card being clickable confirm once)
- [x] Error handling
- [x] Archived Timetables(I assume they lack some features wanted to confirm what they have and what they don't before finishing these)

### CODE REVIEW

File: frontend/src/Home.tsx [hunk 1]
```diff
@@ -128,14 +183,37 @@ function Home() {
...
+              userQueryResult.data?.publicTimetables,
+              false,
+            )}
+
+            {renderTimetableSection(
+              "Draft Timetables:",
+              userQueryResult.data?.draftTimetables,
+              false, // the value of isPrivate shouldn't matter here
+              true,
+            )}
```
Comment (soumitradev) [inline frontend/src/Home.tsx:211]: Here I'd say a few things:
- Show Draft timetables at the top, since we want to emphasize on finishing them first, since the current use case is more focused on making timetables and not as much on sharing them. Private and public you can show in any order that makes sense to you after that.
- If there are no timetables to show in a particular heading then just don't render the heading
- If there are no timetables to show in any of the headings then don't render anything, and instead render a prompt to create a timetable like the current frontend does

File: frontend/src/Home.tsx [hunk 2]
```diff
@@ -128,14 +183,37 @@ function Home() {
...
-      <div className="flex flex-col justify-center items-center pt-20">
-        <h1 className="text-5xl font-bold">There are 2 Routes inside you:</h1>
-        <div className=" pt-16">{JSON.stringify(userQueryResult.data)}</div>
-      </div>
-    </>
-  );
+  if (userQueryResult.isSuccess) {
+    return (
+      <>
+        <main className="bg-primary min-h-screen text-primary-foreground px-8 py-20 md:px-20 lg:px-20 xl:px-48">
```
Comment (soumitradev) [inline frontend/src/Home.tsx:189]: I think the padding-x is too much here in the case of all screens; it should just be a small amount of padding

File: frontend/src/components/TimetableCard.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+          <Button variant="outline">
+            Make {isPrivate ? "Public" : "Private"}
+          </Button>
+        )}
+        <Button variant="ghost">
+          <Edit2 />
+        </Button>
+        <Button variant="ghost">
+          <Trash />
+        </Button>
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:48]: imo make these icons slighly smaller in general for all devices

File: frontend/src/Home.tsx [hunk 4]
```diff
@@ -22,9 +23,63 @@ const fetchUserDetails = async (): Promise<
...
+  title: string,
+  timetables: Timetable[],
+  isPrivate: boolean,
+  isDraft = false,
+) => {
+  return (
+    <>
+      <section className="pt-8 px-4">
+        <h2 className="text-3xl">{title}</h2>
+        <div className="grid gap-8 grid-cols-1 sm:grid-cols-2 lg:grid-cols-3 pt-4">
```
Comment (soumitradev) [inline frontend/src/Home.tsx:56]: imo instead of varying the grid columns, just make a flex container and dump them there with flex-wrap, I feel like that looks better than the variable gap on different devices of different sizes. e.g. The gap on large devices is way too much rn, but the gap on medium to small devices is fine although still on the higher side

tl;dr: reduce gap in general a little bit from 8, and then consider changing this to a flex with flex-wrap instead. Make sure you consult with people on discord if that would look better or not.

File: frontend/src/components/TimetableCard.tsx (hunk 3 above)
Comment (soumitradev) [reply frontend/src/components/TimetableCard.tsx:48]: also if possible make the icon buttons completely round

File: frontend/src/components/TimetableCard.tsx [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+
+type Props = {
+  timetable: z.infer<typeof timetableType>;
+  isPrivate: boolean;
+  isDraft: boolean;
+};
+
+function TimetableCard({ timetable, isPrivate, isDraft }: Props) {
+  return (
+    <Card className="min-w-fit max-w-sm min-h-60 flex flex-col">
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:22]: also this card as of now lacks any kind of hover feedback so fix that too ig

**Review by soumitradev — CHANGES_REQUESTED** [review]
Few changes mostly related to styling of the page

File: frontend/src/components/TimetableCard.tsx (hunk 3 above)
Comment (pnicto) [reply frontend/src/components/TimetableCard.tsx]: I tried making them smaller and I don't like it if others also say the same I will make them smaller.

File: frontend/src/components/TimetableCard.tsx (hunk 5 above)
Comment (pnicto) [reply frontend/src/components/TimetableCard.tsx]: one day, when the card colors make their debut

File: frontend/src/Home.tsx [hunk 6]
```diff
@@ -22,9 +30,65 @@ const fetchUserDetails = async (): Promise<
...
+  return (
+    <>
+      <section className="pt-8">
+        <h2 className="text-3xl font-bold">{title}</h2>
+        <div className="flex flex-wrap gap-8 pt-4">
+          {timetables.length <= 0 ? (
+            <p className="text-xl col-span-full text-center">
+              No timetables to show
+            </p>
+          ) : (
```
Comment (soumitradev) [inline frontend/src/Home.tsx:70]: this part is now redundant, you can edit the condition on line 59 to be `<=` too.

File: frontend/src/components/TimetableCard.tsx [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+              isDraft: true,
+            })
+          }
+        >
+          <Edit2 />
+        </Button>
+        <Button
+          variant="ghost"
+          className="rounded-full p-3"
+          onClick={() => deleteMutation.mutate()}
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:116]: here this should probably display a confirmation modal or something before directly calling the API

File: frontend/src/components/TimetableCard.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+            Make {isPrivate ? "Public" : "Private"}
+          </Button>
+        )}
+        <Button
+          variant="ghost"
+          className="rounded-full p-3"
+          onClick={() =>
+            editMutation.mutate({
+              name: timetable.name,
+              isPrivate: timetable.private,
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:106]: this fails for public timetables, isPrivate should always be false here

File: frontend/src/components/TimetableCard.tsx [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+            <span>|</span>
+            <span className="flex-none">{`${timetable.year}-${timetable.semester}`}</span>
+          </p>
+        </Badge>
+      </CardContent>
+      <CardFooter className="flex justify-end gap-2 mt-auto">
+        {!isDraft && (
+          <Button
+            variant="outline"
+            onClick={() =>
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:89]: This should also probably show a confirmation modal but maybe confirm with others if they think the same, I'm not fully convinced if it should either

File: frontend/src/components/TimetableCard.tsx [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+          title: "Error",
+          description: error.response?.data.message,
+          variant: "destructive",
+        });
+      }
+    },
+  });
+
+  return (
+    <Card className="min-h-60 flex flex-col min-w-80 shadow-lg">
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:70]: probably reduce the min width on smaller devices cuz on some of the phone models the cards are either not centred or are overflowing

File: frontend/src/components/TimetableCard.tsx [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+              isPrivate: timetable.private,
+              isDraft: true,
+            })
+          }
+        >
+          <Edit2 />
+        </Button>
+        <Button
+          variant="ghost"
+          className="rounded-full p-3"
```
Comment (soumitradev) [inline frontend/src/components/TimetableCard.tsx:115]: here this should probably have `hover:bg-destructive/90 hover:text-destructive-foreground` as it's a destructive action, but don't make the entire button destructive cuz that looks ugly

File: frontend/src/Home.tsx [hunk 12]
```diff
@@ -128,14 +215,59 @@ function Home() {
...
+      <>
+        <main className="bg-background min-h-screen text-foreground py-20 px-16">
+          <h1 className="text-5xl font-bold text-center sm:text-left">
+            My Timetables
+          </h1>
+          {draftTimetables.length === 0 &&
+            privateTimetables.length === 0 &&
+            publicTimetables.length === 0 && (
+              <>
+                <div className="bg-secondary mt-10 text-center flex flex-col items-center justify-center gap-8 py-16 rounded-lg">
```
Comment (soumitradev) [inline frontend/src/Home.tsx:232]: here probably a good idea to scale the text and icon according to the device size, it looks too big on phones

File: frontend/src/Home.tsx [hunk 13]
```diff
@@ -128,14 +215,59 @@ function Home() {
...
-    </>
-  );
+  if (userQueryResult.isSuccess) {
+    const { draftTimetables, privateTimetables, publicTimetables } =
+      userQueryResult.data;
+
+    return (
+      <>
+        <main className="bg-background min-h-screen text-foreground py-20 px-16">
+          <h1 className="text-5xl font-bold text-center sm:text-left">
```
Comment (soumitradev) [inline frontend/src/Home.tsx:225]: here the text starts wrapping on phones and it takes up unnecessary vertical space, probably a good idea to reduce the text size on smaller devices

**Review by soumitradev — CHANGES_REQUESTED** [review]
more style related changes

File: frontend/src/Home.tsx [hunk 14]
```diff
@@ -128,14 +273,59 @@ function Home() {
...
-      </div>
-    </>
-  );
+  if (userQueryResult.isSuccess) {
+    const { draftTimetables, privateTimetables, publicTimetables } =
+      userQueryResult.data;
+
+    return (
+      <>
+        <main className="bg-background min-h-screen text-foreground py-20 px-10 md:px-16">
```
Comment (soumitradev) [inline frontend/src/Home.tsx:282]:
```suggestion
        <main className="text-foreground py-6 md:py-12 px-10 md:px-16">
```
Since the navbar is now rendering on this page, the spacing here can be reduced, and since the background will anyways be `bg-background`, the `h-screen` stuff is no longer needed.

**Review by soumitradev — CHANGES_REQUESTED** [review]
One last change before we can merge

**Comment by soumitradev** [comment]
Nevermind, archived timetables are still remaining

**Comment by soumitradev** [comment]
So a couple more things:
- Clicking the create timetable button shouldn't just create the timetable, it should also navigate you to that timetable's edit page directly. As this is only a small change, you can choose to implement this right now or make an issue and leave it for a future PR when the edit timetable page is ready.
- I'm not entirely sure if it makes sense to have an edit button separately on a draft timetable, as the default click behavior should take you to the editor. I can see why an edit button can make this default click action more intuitive, however. So I'd say this is not that important but it is best to discuss before we release. You can choose to make an issue for this as well.
- Archived timetables can be deleted, and their name, and visibility can be edited. You cannot make an archived timetable draft, however. Since archived timetables will probably show up at the bottom of the page, and the visibility of archived timetables isn't a priority for this release, you can choose whether you want to implement this feature or create an issue that we can resolve in some future release.
  - Ideally, I'd say the edit behaviour for archived timetables should be such that on pressing the edit button you get a modal where you can edit the name, and toggle the visibility of the timetable between public and private.
  - The delete behaviour for archived timetables would be the exact same as for any other timetable.

**Comment by pnicto** [comment]
> So a couple more things:
> 
> * Clicking the create timetable button shouldn't just create the timetable, it should also navigate you to that timetable's edit page directly. As this is only a small change, you can choose to implement this right now or make an issue and leave it for a future PR when the edit timetable page is ready.
> * I'm not entirely sure if it makes sense to have an edit button separately on a draft timetable, as the default click behavior should take you to the editor. I can see why an edit button can make this default click action more intuitive, however. So I'd say this is not that important but it is best to discuss before we release. You can choose to make an issue for this as well.
> * Archived timetables can be deleted, and their name, and visibility can be edited. You cannot make an archived timetable draft, however. Since archived timetables will probably show up at the bottom of the page, and the visibility of archived timetables isn't a priority for this release, you can choose whether you want to implement this feature or create an issue that we can resolve in some future release.
>   
>   * Ideally, I'd say the edit behaviour for archived timetables should be such that on pressing the edit button you get a modal where you can edit the name, and toggle the visibility of the timetable between public and private.
>   * The delete behaviour for archived timetables would be the exact same as for any other timetable.

For the 2nd one, I have asked the chat for their opinion as this is a small change. If nothing conclusive comes out then I will make an issue.

File: frontend/src/components/TimetableCard.tsx (hunk 5 above)
Comment (pnicto) [reply frontend/src/components/TimetableCard.tsx]: Made an issue for this #132 and I will mark this as resolved.

**Review by soumitradev — CHANGES_REQUESTED** [review]
One last thing: can you include an icon for public/rpivate timetables when they are archived? It's not exactly clear since they live right next to each other under the subheading.

**Review by soumitradev — APPROVED** [review]
LGTM
